src/services/semver.ts

```diff
@@ -0,0 +1,55 @@
+/* @internal */
+namespace ts {
+    function intOfString(str: string): number {
+        const n = parseInt(str, 10);
+        if (isNaN(n)) {
+            throw new Error(`Error in parseInt(${JSON.stringify(str)})`);
+        }
+        return n;
+    }
+
+    export class Semver {
+        static parse(semver: string): Semver {
+            const isPrerelease = /^(.*)-next.\d+/.test(semver);
+            const result = Semver.tryParse(semver, isPrerelease);
+            if (!result) {
+                throw new Error(`Unexpected semver: ${semver} (isPrerelease: ${isPrerelease})`);
+            }
+            return result;
+        }
+
+        static fromRaw({ major, minor, patch, isPrerelease }: Semver): Semver {
+            return new Semver(major, minor, patch, isPrerelease);
+        }
+
+        // This must parse the output of `versionString`.
+        static tryParse(semver: string, isPrerelease: boolean): Semver | undefined {
+            // Per the semver spec <http://semver.org/#spec-item-2>:
+            // "A normal version number MUST take the form X.Y.Z where X, Y, and Z are non-negative integers, and MUST NOT contain leading zeroes."
+            const rgx = isPrerelease ? /^(\d+)\.(\d+)\.0-next.(\d+)$/ : /^(\d+)\.(\d+)\.(\d+)$/;
+            const match = rgx.exec(semver);
+            return match ? new Semver(intOfString(match[1]), intOfString(match[2]), intOfString(match[3]), isPrerelease) : undefined;
+        }
+
+        constructor(
+            readonly major: number, readonly minor: number, readonly patch: number,
+            /**
+             * If true, this is `major.minor.0-next.patch`.
+             * If false, this is `major.minor.patch`.
+             */
+            readonly isPrerelease: boolean) { }
+
+        get versionString(): string {
+            return this.isPrerelease ? `${this.major}.${this.minor}.0-next.${this.patch}` : `${this.major}.${this.minor}.${this.patch}`;
+        }
+
+        equals(sem: Semver): boolean {
+            return this.major === sem.major && this.minor === sem.minor && this.patch === sem.patch && this.isPrerelease === sem.isPrerelease;
+        }
+
+        greaterThan(sem: Semver): boolean {

```

Comment:
Should `isPrerelease` override anything else in this method?

Reply:
Factored in `isPrerelease`.